### DISCUSSION TITLE
Intersection between a plane and a Manifold

### DISCUSSION THREAD

**Discussion by johnnybigert** [elalish/manifold#437] [Q&A]
Hi!

I would like to find the polygons with holes that constitute the intersection between a plane and a Manifold. I understand that I could use the TrimByPlane function and retrieve the polygons from the resulting Manifold, but it seems wasteful to do a full boolean operation for this. Is there a more efficient way?

Also, I would be really interested in mapping the edges of the intersection back to the original triangles of the Manifold (i.e., for a given edge in the intersection, what triangle in the original Manifold did it originate from). Would that be possible to accomplish?

**Comment by pca006132** [comment]
This should be #426 and is yet implemented.

**Reply by elalish** [reply]
Agreed. Given that we still need to design an API for this, can you give a little more detail about how you want to use these resulting polygons and relations? 

Honestly though, the process of finding this intersection is going to be basically the same as the Boolean op anyway - I might just use it in the implementation. Basically I could just skip the triangulation and simplification. Have you tried to accomplish your goal with our Boolean and `faceID`s?

**Reply by johnnybigert** [reply]
I managed to implement the TrimByPlane idea and retrieved the contours in a Python program without problems.

However, I read the documentation on the faceID vector (in the MeshGL class). It says "contains an ID of the source face this triangle comes from", but given that the TrimByPlane is a boolean operation between a Manifold and a half-space, will the resulting faceIDs then map into the Manifold only? There's probably something here I don't understand, so could you please clarify how the faceIDs work in the TrimByPlane situation?

Also, it seems faceID is not exposed in the Python API, right?

**Reply by elalish** [reply]
Indeed, the triangles on the manifold's side of the polygon will have useful faceIDs corresponding to your input manifold. The triangles on the flat side will all reference some random faceID from the cube used for trimByPlane, so those won't help. However, if you construct a halfedge data structure like we use internally, then it's easy to find the triangle on the other side of the edge and use its ID. I've been considering if it would be helpful to expose this halfedge data structure instead of making users recreate it; I'm open to input here.

Yes, our Python bindings could probably use some love; if you're interested in contributing that would be fantastic!

**Reply by johnnybigert** [reply]
I did a quick and dirty hack to expose a get_face_ids function in Manifold (convert to MeshGL and ask for face ids), just to do the following experiment:

1. I create a Manifold with `numVertices` vertices and `numTriangles` triangles.
2. I use the trim_by_plane function and get a trimmed Manifold with `numTrimmedTriangles`
3. I use the new get_face_ids on the trimmed manifold and get an array of `numTrimmedTriangles` integers in the range `[0, numTriangles]` (all good so far)
4. I check which of the trimmed triangles have two vertices in the plane. (I expect these to originate from some larger, co-planar triangle in the original mesh having been cut by the plane.)
5. I look up the triangles from (4) in the array from (3) to find the triangles from the original Manifold.

However, the result in (5) is not what I expect. I do find co-planar triangles from the original Manifold, but they are never overlapping the trimmed triangle. Why is this you think? Again, there's probably something I don't understand here, what does the face_id of a cut triangle correspond to? (When I apply the reverse mapping from (3) on any other triangle of the trimmed Manifold, I correctly find a triangle with identical vertices from the original mesh, so the mapping itself seems fine.)

Maybe the above is a silly way of doing it, but I just want to see if I can figure out what triangles of the original mesh is being cut by the plane before I start thinking about a better way of doing it :)

**Reply by elalish** [reply]
It's hard to tell from the description, but I think you're generally on the right track. First, they're *face* IDs, so are you interested in triangles specifically, or identifying entire coplanar faces? Second, when you look for vertices "in the plane", are you doing a geometric check for that? You should avoid those at all cost - you can find these triangles topologically by finding shared edges that reference different `originalID`s on each side. 

In answer to your question, I have this line in our docs about `faceID`: 
> When auto-generated, this ID will be a triangle index into the original mesh. All neighboring coplanar triangles from that input mesh will refer to a single triangle of that group as the faceID.

Which triangles is chosen as the ID for a group is arbitrary, and it may not be the one that intersects. However, every output tri that has that ID will be coplanar and some of them will intersect if that original face did at all.

**Reply by johnnybigert** [reply]
Thank you for the clarification!

I assumed that, since I have a triangulated mesh as input, the faceIDs would then refer to the original triangles. But did I understand you correctly, that the algorithm might collapse multiple original, co-planar triangles into a single face, and assign one of the original triangles to represent that face. And it's that representative triangle I see when I call get_face_ids? That would explain things. Is there a way to find the original _triangle_ from an output ("trimmed") triangle somehow?

You mentioned "finding shared edges that reference different originalIDs on each side": I guess I would use MeshGL::runOriginalId for this? And for each neighbouring pair of triangles, check if the originalIDs differ?

Let me explain a bit more about what I'm doing. :) At the moment, I'm just evaluating your library (and it looks _fantastic_) as a candidate to replace the mesh library we're currently using. One of the things I would need to accomplish going forward is, for example, to get the "upper envelope of a Manifold from e.g. Z positive" (which is essentially a Manifold that is the result of projecting all interior points of the original Manifold towards infinite Z negative). When doing this, I would also like to assign an original triangle index to each triangle of the upper envelope Manifold, so that we capture which original triangles are visible from Z positive. Maybe you have some ideas on how to accomplish this in the code base?

**Reply by elalish** [reply]
> You mentioned "finding shared edges that reference different originalIDs on each side": I guess I would use MeshGL::runOriginalId for this? And for each neighbouring pair of triangles, check if the originalIDs differ?

Yes, that's correct. If you want to make sure each triangle gets mapped uniquely via faceID, you can simply pass an extra vertex property, since those also have to be coplanar for triangles to be considered a face. Passing in a random float should be sufficient. 

You probably don't want to project points to infinity (or large numbers), as we'll lose precision and decimate your whole model away. I would probably split the mesh into two halves: triangles with normals pointing up, and those with normals pointing down. Then throw away one half and build vertical faces from all the open edges down to Z=0 and triangulate the bottom polygon(s). Of course this can result in self-intersections in the general case, but you'l have to wait for #289 for help with that. 

Is this for 3D printing support structures? I'd love to know about your software, especially if you do end up using us: #340.

**Reply by johnnybigert** [reply]
Clever with the extra property on the vertices, I will try it out!

We don't have to project to infinity, it's sufficient to stop at the bounding box Z min. No normals in the upper envelope should be pointing down (except the bottom), all normals should be pointing up or be vertical.

I'd love to set up a video call to describe what we're building!

**Reply by elalish** [reply]
You know, I have too many video calls as it is 😁. I work open source, so think I'm going to take a page out of McNeel's book (they make Rhino) and say I'm not keeping anyone's secrets: I won't sign NDAs and if you'd like my opinion on something, let's talk about it publicly in writing. Any startup founder who's worried about someone stealing their idea has not yet figured out what makes success.

**Reply by johnnybigert** [reply]
Of course, no worries. It's just hard to describe without interactivity.

I created this helper in the Python bindings:
```      .def(
          "create_unique_vertices",
          [](Manifold self) {
            Manifold out = self.SetProperties(
              1, [](glm::vec3 pos, const float* oldProp, float* prop) {
                prop[0] = 0.5;  // change to random later
              });
            return out;
          },
          "Creates a new Manifold with unique vertices (so that each triangle"
          "becomes its own faceID).")
```
and run the following Python code in a Jupyter Notebook:
```
manifold_raw = Manifold().from_mesh(mesh)
manifold = manifold_raw.create_unique_vertices()
test = manifold.to_mesh()
trimmed_manifold = manifold.trim_by_plane((0, 0, 1), height)
trimmed_manifold.num_vert()
trimmed_mesh = trimmed_manifold.to_mesh()     # crashes
```
It's interesting that all operations on manifold and trimmed_manifold work out, except the last one. Do you have any ideas why to_mesh would crash? (It all works if I don't run create_unique_vertices, so the crash is related to the new prop.)

**Reply by elalish** [reply]
Hmm, that looks like a bug; we don't have nearly enough testing on our Python bindings, so we may have an error. Also, did you add the setProperties binding? I don't think anyone else has yet...

You can verify it's a binding issue by adding a similar C++ `TEST` and ensuring that passes.

**Reply by johnnybigert** [reply]
I will get back to this in a while and post back here with the results! Meanwhile, I'm including Manifold in some production code scenarios to put it through some real tests ;)
